web-app/HeimScripts/heatmap/run.R

```diff
@@ -176,3 +189,85 @@ add.subset.label <- function(df,label){
   }
   return(df)
 }
+
+computeDendrogram <- function(zScoreMatrix, distanceMeassure, linkageMethod) {
+    as.dendrogram(hclust(dist(zScoreMatrix, method=distanceMeassure), method=linkageMethod))
+}
+
+dendrogramToJSON <- function(d) {
+    totalMembers <- attributes(d)$members
+    add_json <- function(x, start, left) {
+        members <- attributes(x)$members
+        height <- attributes(x)$height
+        index <- (start - 1):(start + members - 2)
+        index <- paste(index, collapse=' ')
+        jsonString <<- paste(jsonString, sprintf('{"height":"%s", "index":"%s", "children":[', height, index))
+        if (is.leaf(x)){
+            jsonString <<- paste(jsonString, ']}')
+        } else {
+            add_json(x[[1]], start, TRUE)
+            jsonString <<- paste(jsonString, ",")
```

Comment:
Yes, this gives an error:

```
f <- function(p) {
  if (p) {
    v <<- 7
  } else {
    p(T)
  }
}
f()
print(v)
```

though here it will probably work, because the variable will be defined one the environement associated with the first call and the environment of the later recursive calls will delegate to this first environment, where the variable already exists, like here:

```
f <- function(p) {
  v <- 5
  if (p) {
    v <<- 7
  } else {
    p(T)
  }
}
f()
print(v)
```

Still, I find it a bit confusing.